enterprise/wsproxy/wsproxysdk/client.go

```diff
@@ -0,0 +1,62 @@
+package wsproxysdk
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+
+	"github.com/coder/coder/coderd/httpmw"
+	"github.com/coder/coder/codersdk"
+)
+
+// Client is a HTTP client for a subset of Coder API routes that external
+// proxies need.
+type Client struct {
+	CoderSDKClient *codersdk.Client
+	// HACK: the issue-signed-app-token requests may issue redirect responses
+	// (which need to be forwarded to the client), so the client we use to make
+	// those requests must ignore redirects.
+	CoderSDKClientIgnoreRedirects *codersdk.Client
+}
+
+// New creates a external proxy client for the provided primary coder server
+// URL.
+func New(serverURL *url.URL) *Client {
+	coderSDKClient := codersdk.New(serverURL)
+	coderSDKClient.SessionTokenHeader = httpmw.WorkspaceProxyAuthTokenHeader
+
+	coderSDKClientIgnoreRedirects := codersdk.New(serverURL)
+	coderSDKClientIgnoreRedirects.HTTPClient.CheckRedirect = func(req *http.Request, via []*http.Request) error {
+		return http.ErrUseLastResponse
+	}
+	coderSDKClientIgnoreRedirects.SessionTokenHeader = httpmw.WorkspaceProxyAuthTokenHeader
+
+	return &Client{
+		CoderSDKClient:                coderSDKClient,
+		CoderSDKClientIgnoreRedirects: coderSDKClientIgnoreRedirects,
+	}
+}
+
+// SetSessionToken sets the session token for the client. An error is returned
+// if the session token is not in the correct format for external proxies.
+func (c *Client) SetSessionToken(token string) error {
+	c.CoderSDKClient.SetSessionToken(token)
+	c.CoderSDKClientIgnoreRedirects.SetSessionToken(token)
+	return nil
+}
+
+// SessionToken returns the currently set token for the client.
+func (c *Client) SessionToken() string {
+	return c.CoderSDKClient.SessionToken()
+}
+
+// Request wraps the underlying codersdk.Client's Request method.
+func (c *Client) Request(ctx context.Context, method, path string, body interface{}, opts ...codersdk.RequestOption) (*http.Response, error) {
+	return c.CoderSDKClient.Request(ctx, method, path, body, opts...)
+}
+
+// RequestIgnoreRedirects wraps the underlying codersdk.Client's Request method
+// on the client that ignores redirects.
+func (c *Client) RequestIgnoreRedirects(ctx context.Context, method, path string, body interface{}, opts ...codersdk.RequestOption) (*http.Response, error) {
+	return c.CoderSDKClientIgnoreRedirects.Request(ctx, method, path, body, opts...)
+}
```

Comment:
I think this is a situation where having a context key that the client uses to ignore redirects in `codersdk` directly could be cleaner. Having two separate clients is jank.

Reply:
We could probably do this, but sadly you can't turn off redirects for a single request which is pretty dumb. We could clone the HTTP client if they specify that in the context but that seems kinda janky too

Reply:
That is kinda janky too. I'm fine with this, but I might hide the `CoderSDKClientIgnoreRedirects` from being exposed. It's a gross (but nicely verbose) thing.

Reply:
I will remove the Coder prefixes.